lib/AuthStateManager.ts

```diff
@@ -0,0 +1,129 @@
+import { AuthSdkError } from './errors';
+import PromiseQueue from './PromiseQueue';
+import { omit } from './util';
+import { 
+  ACCESS_TOKEN_STORAGE_KEY, 
+  ID_TOKEN_STORAGE_KEY 
+} from './constants';
+import { 
+  AuthState,
+  Tokens, 
+  AccessToken, 
+  IDToken,
+  UpdateAuthStateOptions
+} from './types';
+import { OktaAuth } from './browser';
+
+const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+
+class AuthStateManager {
+  private sdk: OktaAuth;
+  private pending: { tokens: Tokens };
+  private authStateQueue: PromiseQueue;
+  private authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this.sdk = sdk;
+    this.pending = { tokens: {} };
+    this.authStateQueue = new PromiseQueue();
+    this.authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to sync update tokens in memory (this.pending)
+    // And push async update authState and emit process into PromiseQ
+
+    // "added" event is emitted in both add and renew process
+    // Only listen on "added" (instead of both "added" and "renewed") to limit authState re-evaluation
+    sdk.tokenManager.on('added', (key, token) => {
+      this.pending.tokens = { ...this.pending.tokens, [key]: token };
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+    sdk.tokenManager.on('removed', (key) => {
+      this.pending.tokens = omit(this.pending.tokens, key);
+      this.authStateQueue.push(
+        this.updateAuthState, 
+        this,
+        { ...this.pending.tokens }
+      );
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this.authState;
+  }
+
+  async initialAuthState(): Promise<void> {
+    let accessToken = await this.sdk.tokenManager.get(ACCESS_TOKEN_STORAGE_KEY) as AccessToken;
```

Comment:
seems like all this logic should go in `updateAuthState` (and/or the default `isAuthenticated`). The initial authState should run the same calculations we would do on any other authState cycle. Is there something special about the first one?

Reply:
same here: https://github.com/okta/okta-auth-js/pull/453#discussion_r475853065